### DISCUSSION TITLE
[1.1.433] Multi-Inputs unit

### DISCUSSION THREAD

**Discussion by huchenlei** [Mikubill/sd-webui-controlnet#2527] [General]
Related discussion: https://github.com/Mikubill/sd-webui-controlnet/issues/2478#issuecomment-1894173725.
Original request: https://github.com/Mikubill/sd-webui-controlnet/issues/2245.

User has requested that we provide a way to easily input the whole directory of images into a unit. This PR partially does that by providing the `Multi-Inputs` tab. Now you can specify a directory with N images. N ControlNet units will be added on generation each unit accepting 1 image from the dir.

The feature can be very useful on IPAdapter units, as we can create "instant LoRA" with multiple input images from a directory. Previously, you would need to enable multiple ControlNet units, and upload images one by one. This also paves the way for implementing IPAdapter portrait: https://github.com/tencent-ailab/IP-Adapter/commit/2397f7c3735c89cd2c9da3229d95d99673f36464.

Note:
You do not need to increase ControlNet unit count to be greater than the number of images in your directory.
Merging 4 images with IPAdapter (weight=0.2), while only have 3 ControlNet unit in settings.
![1705807774664](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/1f9461b1-1af7-4a2f-8c82-2cfb33420edb)
![1705809716391](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/e9488e13-3f08-4903-b1c2-71671770478b)

## Update (2024/01/21)
A more user friendly UI is implemented in #2533
![1705878420439](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/ffd04040-dba3-4d1e-863e-affa4d6d604b)

**Comment by IndigoDosSantos** [comment]
Very cool, tyvm!

> This also paves the way for implementing IPAdapter portrait: [tencent-ailab/IP-Adapter@2397f7c](https://github.com/tencent-ailab/IP-Adapter/commit/2397f7c3735c89cd2c9da3229d95d99673f36464).

Oh that makes sense, that's why portrait is not yet working?

**Comment by 2blackbar** [comment]
This is another big step for controlnet IMO as quite a lot of models now can support batch input for images.

**Comment by 2blackbar** [comment]
Actually i tried it and it doesnt work as in comfy, the more images you feed it the more it stacks up and you get distorted noise image , with 10 images you already get noise, in comfy i can have even 20 pics with no issues
Also the fact that it loads all images over and over everytime you generate is not ideal, should be just once unless you change path or somethin.
Theres batch images workflow for comfy and this 
 repo probably has info in the code how it handles multiple images https://github.com/cubiq/ComfyUI_IPAdapter_plus/blob/main/examples/IPAdapter_batch_images.json

**Reply by huchenlei** [reply]
Thanks for sharing that workflow. The current multi-input is essentially stacking units with same setting but different inputs together. However, there might be bugs in our impl. I will need some time to investigate.

**Reply by 2blackbar** [reply]
yeah i have no idea how they avoid distortion(image turning into noise like high values with lora) on multiple inputs

**Reply by dill-shower** [reply]
> with 10 images you already get noise, in comfy i can have even 20 pics with no issues.

Can you try set cfg to 10-25? It help for me

**Reply by 2blackbar** [reply]
I can but this should be handled by the code internally like in comfy so people dnt have to memorize all the tricks to make it work

**Comment by wardensc2** [comment]
Excuse me, is this problem fixed ? I can't wait to make an instance Lora with FaceID v2 with multi images input

**Reply by huchenlei** [reply]
Yes. The multi-inputs tab is available in original A1111 sd-webui-controlnet now. The forge PR is waiting to be merged here: https://github.com/lllyasviel/stable-diffusion-webui-forge/pull/264

**Reply by wardensc2** [reply]
Thank you so much

**Comment by snowcoffee8023** [comment]
hi, how to use multi-input in api?

**Reply by redwolf3245** [reply]
Any answer for that?

**Comment by gohan2091** [comment]
Is the Multi-input tab not available in Forge? I have batch folder and batch upload but no mutli-input. Can anyone help?

**Reply by IndigoDosSantos** [reply]
I thought multi-input got renamed to batch upload, but after some quick tests, turns out multi-input is actually missing in Forge.

**Reply by huchenlei** [reply]
https://github.com/lllyasviel/stable-diffusion-webui-forge/pull/264

**Reply by gohan2091** [reply]
> [lllyasviel/stable-diffusion-webui-forge#264](https://github.com/lllyasviel/stable-diffusion-webui-forge/pull/264)

Pointless updating. It adds multi input but breaks ControlNet and IPAdapters....

**Comment by Mr-Qiu97** [comment]
hi, how to use multi-input in api?
